metron-platform/metron-solr/src/main/java/org/apache/metron/solr/dao/SolrMetaAlertDao.java

```diff
@@ -144,10 +144,11 @@ protected String getDefaultSourceTypeField() {
       }
     };
 
-    SolrClient solrClient = solrDao.getSolrClient(solrDao.getZkHosts());
+    SolrClient solrClient = solrDao.getSolrClient();
```

Comment:
I'm still confused as to why we are passing the solr client around in this fashion. I thought Spring is smart enough to create singletons so that all the DAO constructors can take a SolrClient arg. See note on https://github.com/apache/metron/pull/1367/files#r269692863

Reply:
Well, as the original implementer, the technical reasoning is that I didn't know better. There's no compelling reason to leave it this way.

Reply:
The `SolrDao` and `SolrMetaAlertDao` are created separately through reflection in the REST `IndexConfig` class.  Solr is a runtime dependency so the `SolrClient` instance cannot be created in REST.  There must be some way for the client to be shared between them.  The `getSolrClient` is simply returning an existing client (not creating anything) so I feel it's pretty safe.  I'm happy to move it to the factory class if you think that makes more sense.

Reply:
Per @justinleet's comments, I think we should get rid of `getSolrClient()` from the SolrDao. It's not "dangerous" specifically in this case, but it makes it harder for someone else to come along later and understand where the client connections are coming from, how they're being created, and how many there will be. We just need to share a single instance at construction time, or per the issue you cited regarding runtime deps, inject singletons.

The real solution here is to refactor this using traditional GOF patterns and simply use the abstract factory pattern as that's exactly its purpose - https://en.wikipedia.org/wiki/Abstract_factory_pattern. I think it's arguably out of scope for a bug fix if you don't want to do it, but that's my cents on what the best approach would be. Mind you, the DAO combining part might need some thought, but I suspect we can simplify it just fine.

Reply:
The latest commit removes the `getSolrClient` method from the `SolrDao` class.  The SolrClient instance is now managed by the `SolrClientFactory` class.  I had to also add a `close` method to close the connection and release the reference so that a new SolrClient can be instantiated as needed.  This was necessary for integration tests because multiple in memory Solr servers are spun up in the same JVM and a new client is needed for each.